### PR TITLE
How to split code?

### PR DESCRIPTION
**Pull request by pksjce**
Trying to tackle #3 with this.
- [x] Intro to code splitting
- [x] On demand code splitting
   - [x] With require.ensure()
- [x] Resource splitting
   - [x] CSS splitting
   - [x] Vendor chunk splitting (Common chunks and manifest)

### CODE REVIEW

File: content/how-to/code-splitting/splitting-css.md [hunk 1]
```diff
@@ -0,0 +1,83 @@
+---
+title: How to Split CSS?
+contributors:
+  - pksjce
+---
+
+In `webpack`, when you use the css loader and import css into your javascript files, the css is bundled along with your javascript.
```
Comment (bebraw) [inline content/how-to/code-splitting/splitting-css.md]: * Just webpack to be consistent with other docs.
* javascript -> JavaScript
* css -> CSS

File: content/how-to/code-splitting/splitting-css.md [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+This has the disadvantage that, you will not be able to utilize the browser's ability to load css asynchronously and parallely. Instead, your page will have to wait until your whole javascript bundle is loaded, to style itself.
+`webpack` can help with this problem by bundling the css separately using [Extract-Text-Webpack-Plugin](https://github.com/webpack/extract-text-webpack-plugin) and the [css-loader](https://github.com/webpack/css-loader)
+
+## Using `css-loader`
+
+To import css into your javascript code like [any other module](concept/modules), you will have to use the [css-loader](https://github.com/webpack/css-loader)
```
Comment (bebraw) [inline content/how-to/code-splitting/splitting-css.md]: /s/javascript/JavaScript

File: content/how-to/code-splitting/splitting-css.md [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+```
+
+To use this ExtractPlugin, it needs to be added to the `webpack.config.js` file in two steps.
+### In the loader
+
+Adapting from the previous example with the `css-loader`, we should add ExtractPlugin as follows
+
+```javascript
+...
+loader: ExtractPlugin.extract('css-loader?sourceMap') //Can be used without sourcemaps too.
```
Comment (bebraw) [inline content/how-to/code-splitting/splitting-css.md]: The syntax has changed in webpack 2.

File: content/how-to/code-splitting/splitting-css.md [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+...
+```
+
+### In the plugin
+
+```javascript
+new ExtractPlugin({ filename: 'bundle.css', disable: false, allChunks: true })
```
Comment (bebraw) [inline content/how-to/code-splitting/splitting-css.md]: Why `ExtractPlugin` instead of `ExtractTextPlugin`?

File: content/how-to/code-splitting/splitting-css.md (hunk 3 above)
Comment (bebraw) [inline content/how-to/code-splitting/splitting-css.md]: Yup. It's a breaking change.

File: content/how-to/code-splitting/splitting-css.md [hunk 5]
```diff
@@ -0,0 +1,86 @@
+---
+title: How to Split CSS?
+contributors:
+  - pksjce
+---
+
+In webpack, when you use the css-loader and import CSS into your Javascript files, the CSS is bundled along with your Javascript.
+This has the disadvantage that, you will not be able to utilize the browser's ability to load CSS asynchronously and parallely. Instead, your page will have to wait until your whole Javascript bundle is loaded, to style itself.
+webpack can help with this problem by bundling the CSS separately using [Extract-Text-Webpack-Plugin](https://github.com/webpack/extract-text-webpack-plugin) and the [css-loader](https://github.com/webpack/css-loader)
```
Comment (SpaceK33z) [inline content/how-to/code-splitting/splitting-css.md]: Forgot a dot in the end.

File: content/how-to/split-code.md [hunk 6]
```diff
@@ -1,11 +1,30 @@
...
-> what is code splitting
-> System.import
-> dynamic System.import
-> require.ensure
-> AMD require
-
-> see also [[Output]]
-> see also [[dynamic dependencies]]
+
+Code splitting is the most compelling feature for  `webpack` usage. `webpack`, with some configuration, can tweak your app such that, you can split your code into various bundles and load them on demand.
```
Comment (SpaceK33z) [inline content/how-to/split-code.md]: Two spaces before `webpack`.

Comment (SpaceK33z) [inline content/how-to/split-code.md]: Can you take another lok at the second sentence? It has too many comma's in it, making it a bit hard to read.

File: content/how-to/split-code.md [hunk 7]
```diff
@@ -1,11 +1,30 @@
...
+#### Code splitting with require.ensure
+#### Code splitting with System.import
+#### Making Common chunks
+#### Code splitting using bundle loader
+
+## Resource splitting for cacheing and parallel loads
+
+#### CSS splitting
+
+An application owner would want to split all the css into a separate bundle. This enhances caheability of the resource bundle and also allows the browser to parallely load the bundle which makes for a solid performance improvement.
```
Comment (SpaceK33z) [inline content/how-to/split-code.md]: caheability -> cacheability

File: content/how-to/code-splitting/splitting-css.md [hunk 8]
```diff
@@ -0,0 +1,86 @@
+---
+title: How to Split CSS?
+contributors:
+  - pksjce
+---
+
+In webpack, when you use the css-loader and import CSS into your Javascript files, the CSS is bundled along with your Javascript.
```
Comment (bebraw) [inline content/how-to/code-splitting/splitting-css.md]: /s/Javascript/JavaScript

File: content/how-to/code-splitting/splitting-css.md [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+This has the disadvantage that, you will not be able to utilize the browser's ability to load CSS asynchronously and parallely. Instead, your page will have to wait until your whole Javascript bundle is loaded, to style itself.
```
Comment (bebraw) [inline content/how-to/code-splitting/splitting-css.md]: /s/parallely/parallel

Comment (bebraw) [inline content/how-to/code-splitting/splitting-css.md]: /s/Javascript/JavaScript

File: content/how-to/code-splitting/splitting-css.md [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+webpack can help with this problem by bundling the CSS separately using [Extract-Text-Webpack-Plugin](https://github.com/webpack/extract-text-webpack-plugin) and the [css-loader](https://github.com/webpack/css-loader).
```
Comment (bebraw) [inline content/how-to/code-splitting/splitting-css.md]: Maybe lower case extract-text-webpack-plugin.

File: content/how-to/code-splitting/splitting-css.md [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+
+## Using `css-loader`
+
+To import css into your Javascript code like [any other module](concept/modules), you will have to use the [css-loader](https://github.com/webpack/css-loader)
```
Comment (bebraw) [inline content/how-to/code-splitting/splitting-css.md]: JavaScript

File: content/how-to/code-splitting/splitting-css.md [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+The webpack config with `css-loader` will look somewhat like
```
Comment (bebraw) [inline content/how-to/code-splitting/splitting-css.md]: I would rephrase this and drop "somewhat".

File: content/how-to/code-splitting/splitting-css.md [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+
+Install this plugin as follows
+```
+npm i --save-dev extract-text-webpack-plugin
+```
+
+To use this ExtractTextPlugin, it needs to be added to the `webpack.config.js` file in two steps.
+### In the loader
+
+Adapting from the previous example with the `css-loader`, we should add ExtractTextPlugin as follows
```
Comment (bebraw) [inline content/how-to/code-splitting/splitting-css.md]: ExtractTextPlugin is missing formatting.

File: content/how-to/code-splitting/splitting-css.md [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+### In the plugin
+
+```javascript
+new ExtractTextPlugin({ filename: 'bundle.css', disable: false, allChunks: true })
+```
+
+With above two steps, you can generate a new bundle specifically for all the CSS modules and add them as a separate tag in the `index.html`
+For more info on how to use the api please go to [ExtractTextPlugin api](https://github.com/webpack/extract-text-webpack-plugin#api).
+
+The full config for splitting css with ExtractTextPlugin is as follows
```
Comment (bebraw) [inline content/how-to/code-splitting/splitting-css.md]: ExtractTextPlugin formatting.

File: content/how-to/code-splitting/splitting-require.md [hunk 15]
```diff
@@ -0,0 +1,97 @@
+---
+title: How to Split Using require.ensure()?
+contributors:
+  - pksjce
+---
+
+In this section, we will discuss how webpack splits code using `require.ensure()`.
+
+### `require.ensure()`
```
Comment (bebraw) [inline content/how-to/code-splitting/splitting-require.md]: Decrease header level by one. So you get ## here. See also headers below.

File: content/how-to/code-splitting/splitting-require.md [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+
+webpack statically parses for `require.ensure()` in the code while building and adds the modules here into a separate chunk. This new chunk is loaded on demand by webpack through jsonp.
+
+The syntax is as follows
+
+``` javascript
```
Comment (bebraw) [inline content/how-to/code-splitting/splitting-require.md]: Extra whitespace.

File: content/how-to/code-splitting/splitting-require.md [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+
+#### callback
+This is the callback function that webpack will execute once the dependencies are loaded. An implementation of the require object is sent as a parameter to this function. This is so that, we can further `require()` the dependencies and any other modules for execution.
+
+#### chunkName
+The chunkName is the name given to the chunk created by this particular `require.ensure()`. By giving the same name at various different split points of `require.ensure()`, we can make sure all the dependencies are collectively put in the same bundle.
+
+Let us consider the following project
+
+```
```
Comment (bebraw) [inline content/how-to/code-splitting/splitting-require.md]: Maybe looks better with `bash`.

File: content/how-to/code-splitting/splitting-require.md [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+    }
+}
+```
+On running webpack on this project, we find that webpack has created two new bundles, `bundle.js` and `0.bundle.js`.
+
+`entry.js` and `a.js` are bundled in `bundle.js`.
+
+`b.js` is bundled in `0.bundle.js`.
+
+### Gotchas for require.ensure()
```
Comment (bebraw) [inline content/how-to/code-splitting/splitting-require.md]: Missing formatting for require.ensure.

File: content/how-to/code-splitting/splitting-require.md [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+
+#### empty array as parameter
```
Comment (bebraw) [inline content/how-to/code-splitting/splitting-require.md]: Titlecase.

File: content/how-to/code-splitting/splitting-require.md [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
+
+```javascript
+require.ensure([], function(require){
+    require('./a.js')
+})
+```
+
+The above code ensures that a split point is created and `a.js` is bundled separately by webpack.
+
+#### dependencies as parameter
```
Comment (bebraw) [inline content/how-to/code-splitting/splitting-require.md]: Titlecase (Dependencies as Parameter).

File: content/how-to/code-splitting/splitting-require.md [hunk 21]
```diff
@@ -0,0 +1,97 @@
...
+
+```javascript
+require.ensure(['./a.js'], function(require) {
+    require('./b.js');
+})
+```
+
+In the above code, `a.js` and `b.js` are bundled together and split from the main bundle. But only the contents of `b.js` are executed. The contents of `a.js` are only made available and not executed.
+To execute `a.js`, we will have to require it in a sync manner like `require('./a.js')` for the Javascript to get executed.
```
Comment (bebraw) [inline content/how-to/code-splitting/splitting-require.md]: JavaScript

File: content/how-to/code-splitting/splitting-vendor.md [hunk 22]
```diff
@@ -0,0 +1,132 @@
+---
+title: How to Split Vendor Files?
+contributors:
+  - pksjce
+---
+
+A typical application uses third party libraries for framework/functionality needs. Particular versions of these libraries are used and code here does not change often. However, the application code changes frequently. Bundling application code with third party code would be inefficient. This is because the browser can cache asset files based on the cache header and files can be cached without needing to call the cdn again if it's contents don't change. To take advantage of this, we want to keep the hash of the vendor files constant regardless of application code changes. We can do this only when we separate the bundles for vendor and application code.
```
Comment (bebraw) [inline content/how-to/code-splitting/splitting-vendor.md]: Split up into two, three paragraphs.

File: content/how-to/code-splitting/splitting-vendor.md [hunk 23]
```diff
@@ -0,0 +1,132 @@
...
+        }
+    }
+}
+```
+
+On running `webpack` in your application, if you inspect the resulting bundle, you will see that `moment` and `index.js` have been bundled in `bundle.js`.
+
+This is not ideal for the application. If the code in `index.js` changes, then the whole bundle is rebuilt. The browser will have to load a new copy of the new bundle even though most of it hasn't changed at all.
+
+### Multiple entries
```
Comment (bebraw) [inline content/how-to/code-splitting/splitting-vendor.md]: Decrease to ##. See also the rest.

File: content/how-to/code-splitting/splitting-vendor.md [hunk 24]
```diff
@@ -0,0 +1,132 @@
...
+        }
+    }
+}
+```
+
+On running `webpack` now, we see that two bundles have been created. If you inspect these though, you will find that the code for `moment` is present in both the files!
+
+It is for this reason, that we will need to use the [CommonsChunkPlugin](https://github.com/webpack/webpack/blob/master/lib/optimize/CommonsChunkPlugin.js).
+
+### CommonsChunkPlugin
```
Comment (bebraw) [inline content/how-to/code-splitting/splitting-vendor.md]: Header level + missing formatting.

File: content/how-to/code-splitting/splitting-vendor.md [hunk 25]
```diff
@@ -0,0 +1,132 @@
...
+            filename: '[chunkhash].[name].js`,
+            path: './dist'
+        }
+    }
+}
+```
+
+On running `webpack` now, we see that two bundles have been created. If you inspect these though, you will find that the code for `moment` is present in both the files!
+
+It is for this reason, that we will need to use the [CommonsChunkPlugin](https://github.com/webpack/webpack/blob/master/lib/optimize/CommonsChunkPlugin.js).
```
Comment (bebraw) [inline content/how-to/code-splitting/splitting-vendor.md]: This link can break if the source changes. Maybe link to a hash (specific commit). Not sure.

File: content/how-to/code-splitting/splitting-vendor.md [hunk 26]
```diff
@@ -0,0 +1,132 @@
...
+            new webpack.optimize.CommonsChunkPlugin({
+                name: 'vendor' // Specify the common bundle's name.
+            })
+        ]
+    }
+}
+```
+Now run `webpack` on your application. Bundle inspection shows that `moment` code is present only in the vendor bundle.
+
+### Manifest file
```
Comment (bebraw) [inline content/how-to/code-splitting/splitting-vendor.md]: Manifest File. See also the rest.

File: content/how-to/code-splitting/splitting-vendor.md [hunk 27]
```diff
@@ -0,0 +1,132 @@
...
+
+But, if we change application code and run `webpack` again, we see that the hash for the vendor file changes. Even though we achieved separate bundles for `vendor` and `main` bundles, we see that the `vendor` bundle changes when the application code changes.
+This means that we still don't reap the benefits of browser caching because the hash for vendor file changes on every build and the browser will have to reload the file.
+
+The issue here is that on every build, webpack generates some webpack runtime code, which helps webpack do it's job. When there is a single bundle, the runtime code resides in it. But when multiple bundles are generated, the runtime code is extracted into the common module, here the `vendor` file.
+
+To prevent this, we need extract out the runtime into a separate manifest file. Even though we are creating another bundle, the overhead is offset by the long term caching benefits that we obtain on the `vendor` file.
+
+```javascript
+\\ webpack.config.js
```
Comment (bebraw) [inline content/how-to/code-splitting/splitting-vendor.md]: Push filenames outside of the examples so you get something like this:

**webpack.config.js**

```javascript
demo goes here
```

File: content/how-to/split-code.md [hunk 28]
```diff
@@ -1,11 +1,35 @@
...
+
+## On demand code-splitting
+
+`webpack` can help us split our code into logical pieces or chunks as per our application routes or as per predicted user behaviour. This means that we can load non-essential assets when the user performs an action like route change and demands for it.
+
+### Code splitting with `require.ensure()`
+
+`require.ensure()` is the CommonJS way of including assets asynchronously. By adding `require.ensure([<fileurl>])`, we can define a split point in the code. webpack can then create a separate bundle of all the code inside this split point.
+Learn [how to split your code using `require.ensure()`](/how-to/code-splitting/splitting-require)
+### Code splitting with System.import
```
Comment (bebraw) [inline content/how-to/split-code.md]: Add TODOs (?>) here so we won't forget to document these.

File: content/how-to/code-splitting/splitting-vendor.md (hunk 25 above)
Comment (pksjce) [inline content/how-to/code-splitting/splitting-vendor.md]: Not sure what exactly to do. Shall I remove the link?

Comment (bebraw) [inline content/how-to/code-splitting/splitting-vendor.md]: You could add `plugins/commons-chunk-plugin.md` stub and link to it instead.

**Comment by bebraw** [comment]
Can you rebase? I'll merge after that.

**Comment by pksjce** [comment]
Done!
